healthz/README.md

```diff
@@ -7,41 +14,243 @@ in distributed systems.
+artifacts that can be returned by the system.
+
+All artifacts that are listed within the same `ComponentStatus` message are
+expected to have a shared acknowledgement state, and expiry time.
+
+### Healthz.List()
+
+```protobuf
+rpc List(ListRequest) returns (ListResponse) {}
+```
+
+`List` returns all events that are associated with a particular component. In
+contrast to `Get`, it returns all events that the target has for the device. By
+default, events that are acknowledged are not returned.
+
+Similarly to `Get`, a series of `ComponentStatus` messages are returned with
+the same semantics as are described above.
+
+### Healthz.Acknowledge()
+
+```protobuf
+rpc Acknowledge(AcknowledgeRequest) returns (AcknowledgeResponse) {}
+```
+
+`Acknowledge` is used by a client to indicate to the target device that a
+particular (component, event) tuple has been retrieved by the client. This
+allows a device to intelligently determine whether to retain artifacts. Devices
+MUST ensure that artifact storage for healthz does not cause resource exhaustion
+and SHOULD remove acknowledged artifacts before those that have not yet received
+an acknowledgement.
+
+### Healthz.Artifact() 
+
+```protobuf
+rpc Artifact(ArtifactRequest) returns (stream ArtifactResponse) {}
 ```
 
-## Use Cases
+`Artifact` allows a user to retrieve a specific artifact that is related with
+an event that has occurred. Since these artifacts may be large, the `Artifact`
+RPC is implemented as a server-side streaming RPC. Use of the `Artifact` RPC
+ensures that the resources to send these potentially large artifacts only
+when explicitly requested by the client.
+
+#### Healthz.Check()
+
+```protobuf
+rpc Check(CheckRequest) returns (CheckResponse) {}
+```
+
+`Check` allows a client to execute a set of "validations" against the specified
+component. The component, as with other operations, is specified in terms of
+the gNMI path.  
+
+The result of the `Check` produces a healthz `ComponentStatus` message which
+will contain a list of the generated artifacts used in the validation process.
+
+
+A call to the `Check` endpoint will allow for "very expensive" debugging
+commands such as causing the device to pause and snapshot its system database
+or to examine the state of a protocol.  These commands will likely be
+considered "service impacting" and should have a clear security ACL restricting
+their use during normal operations of the device.
+
+## User Experience
+
+#### A BGP routing process goes unhealthy due to a crash.
+
+##### Expected Action
+
+Write the core dump to a file system.
+
+A call to `Healthz.Get` should respond with the current state of the component
+and provide feedback that there is an artifact available to be requested.
+
+A Healthz call should stream the core dump over the service back to the caller.
+
+The component that is specified should correspond to the software process that
+implements the BGP protocol in the system.
```

Comment:
I think it would be very cool to augment this process with a few more details, which would clarify the interopretations between healthz service and telemetry paths.

1. shall we add here a step after the core dump is written that goes like: 
    > A Healthz event MUST be generated stating the new health state is UNHEALTHY. It is left for the implementor to decide which artifacts are generated during this automatic event generation following the rule that those automatic event generations SHOULD NOT affect the performance of a system. If the triage of this failure requires substantial computational resource, `Healthz.Check()` must be used by the client.
2. Clarify, that `Healthz.Get()` might not list (all?) the artifacts, if collection of those artifacts was considered to be resource intensive task. In such case, the `Check()` RPC should be used to populate the list of available artifacts.
3. Add a note, that telemetry paths are updated automatically, without calling any healthz RPCs denoting the changed health state

Reply:
Hi @robshakir 
can you ptal at this comment? The rest LGTM (with a small term glitch [here](https://github.com/openconfig/gnoi/pull/106#discussion_r1012733954))

Reply:
I think this makes sense -- from a "safe" implementation perspective. Just looking at the place to add this to the doc.

Reply:
Added a paragraph that basically says exactly what you say further up in the document since I think it is important for the designers of clients too.